crud/stats/module.lua

```diff
@@ -0,0 +1,211 @@
+local clock = require('clock')
+local checks = require('checks')
+local errors = require('errors')
+local vshard = require('vshard')
+
+local dev_checks = require('crud.common.dev_checks')
+local utils = require('crud.common.utils')
+local op_module = require('crud.stats.operation')
+local registry = require('crud.stats.local_registry')
+
+local StatsError = errors.new_class('StatsError', {capture_stack = false})
+
+local stats = {}
+local is_enabled = false
+
+--- Initializes statistics registry, enables callbacks and wrappers
+--
+--  If already enabled, do nothing.
+--
+-- @function enable
+--
+-- @treturn boolean Returns true.
+--
+function stats.enable()
+    if is_enabled then
+        return true
+    end
+
+    StatsError:assert(
+        rawget(_G, 'crud') ~= nil,
+        "Can be enabled only on crud router"
+    )
+
+    registry.init()
+    is_enabled = true
+
+    return true
+end
+
+--- Resets statistics registry
+--
+--  After reset collectors set is the same as right
+--  after first stats.enable().
+--
+-- @function reset
+--
+-- @treturn boolean Returns true.
+--
+function stats.reset()
+    if not is_enabled then
+        return true
+    end
+
+    registry.destroy()
+    registry.init()
+
+    return true
+end
+
+--- Destroys statistics registry and disable callbacks
+--
+--  If already disabled, do nothing.
+--
+-- @function disable
+--
+-- @treturn boolean Returns true.
+--
+function stats.disable()
+    if not is_enabled then
+        return true
+    end
+
+    registry.destroy()
+    is_enabled = false
+
+    return true
+end
+
+--- Get statistics on CRUD operations
+--
+-- @function get
+--
+-- @tparam string space_name
+--  (Optional) If specified, returns table with statistics
+--  of operations on space, separated by operation type and
+--  execution status. If there wasn't any requests of "op" type
+--  for space, there won't be corresponding collectors.
+--  If not specified, returns table with statistics
+--  about all observed spaces and count of calls to spaces
+--  that wasn't found.
+--
+-- @treturn table Statistics on CRUD operations.
+--  If statistics disabled, returns {}.
+--
+function stats.get(space_name)
+    checks('?string')
+
+    return registry.get(space_name)
+end
+
+local function wrap_tail(space_name, op, opts, start_time, call_status, ...)
+    local finish_time = clock.monotonic()
+    local latency = finish_time - start_time
+
+    local err = nil
+    local status = 'ok'
+    if call_status == false then
+        status = 'error'
+        err = select(1, ...)
```

Comment:
err = select(1, ...) is the same as `err = ...` but without select.

Reply:
But this one is more readable and makes code consistent. Does it adds any overhead?

Reply:
To be honest I don't know. I only could to compare amount of bytecode that produces each command (note! I test using LuaJIT 2.0.5, and it could be different on LuaJIT 2.1-beta3).

```
cat test.lua 
variable = ...

luajit -bl test.lua
-- BYTECODE -- test.lua:0-2
0001    VARG     0   2   0
0002    GSET     0   0      ; "variable"
0003    RET0     0   1
```

```
cat test.lua 
variable = select(1, ...)

- BYTECODE -- test.lua:0-2
0001    GGET     0   0      ; "select"
0002    KSHORT   1   1
0003    VARG     2   0   0
0004    CALLM    0   2   1
0005    GSET     0   1      ; "variable"
0006    RET0     0   1
```